### PR TITLE
Remove RequestDelegateFactory call from RouteEndpointBuilder

### LINKED ISSUE #42703: Remove RequestDelegateFactory call from RouteEndpointBuilder
The proposal is to remove the call to `RequestDelegateFactory.Create()` I added to `RouteEndpointBuilder.Build()` in #42195. The discussion surrounding its inclusion is [here](https://github.com/dotnet/aspnetcore/pull/42195#discussion_r901049629). The main impact of this change is that filters run on all endpoints, not just minimal route handlers.

Currently, if an endpoint is not a minimal route handler, the method being filtered is just a `RequestDelegate` even if that `RequestDelegate` ultimately called an MVC action or SignalR Hub method with a different method signature.

If we only stopped calling `RDF.Create()` in `REB.Build()` and made no other changes, this would mean filters would go back to only working with minimal route handlers despite now being applicable to any `IEndpointConventionBuilder`. We could conceivably come up with analyzers that give errors when applied to `IEndpointConventionBuilder`s we know won't work, but that still doesn't help that filters applied to groups would only run for minimal route handlers and no other endpoints.

Fortunately, in addition to removing `RDF.Create()` from `REB.Build()`, we're planning to add better support for what are now "EndpointFilters" instead of "RouteHandlerFilters" in MVC with #42557. With this PR, the actual MVC action ends the filter pipeline instead of a `RequestDelegate` that indirectly invokes the action.

Furthermore, we should update our internal `ModelEndpointDataSource` to run filters on the `RequestDelegate`s passed to `MapGet` and its associated methods. It make sense to use `RequestDelegate` as the final method in the filter pipeline in this case, because that's what the developer defined the endpoint with.

Some things like `MapHub` just call the `RequestDelegate` `Map` method overloads. For things like SignalR Hubs that might have another plausible way of filtering the endpoint other than just by filtering the `RequestDelegate` it defines, we should probably start using a custom internal `EndpointDataSource` that does not run filters on the `RequestDelegate`s.

Based on early discussions, we don't think we'll be running these filters for Razor Pages or SignalR Hub invocations in .NET 7. If we decide to run filters on more things in .NET 8, we'll need to figure out how to do it in a way that is sufficiently non-breaking way, but we'll cross that bridge when we get there. Any change to filter behavior without an opt-in switch is potentially breaking.

In general, the change here is to make the `EndpointDataSource` implementations solely responsible for running filters. This means that anyone using a custom `EndpointDataSource` won't have filters run automatically when building their `RouteEndpoint`s, but at least they can run them manually. I think custom `EndpointDataSource` implementations are relatively rare in any case.

### PR DESCRIPTION
**Pull request by halter73**
- Remove ModelEndpointDataSource. Use RouteEndpointDataSource instead
- Filters will no longer run for razor pages due to this change.

This is somewhat dependent on #42557 to ensure we have a nice experience where "endpoint" filters run on both route handlers, normal RequestDelegates and MVC actions.

Fixes #42703

### CODE REVIEW

File: src/Http/Routing/src/RouteEndpointDataSource.cs
Comment (davidfowl) [inline src/Http/Routing/src/RouteEndpointDataSource.cs:167]: I think we want to add this in either case so open api works.

File: src/Http/Routing/test/UnitTests/RouteEndpointBuilderTest.cs
Comment (davidfowl) [inline src/Http/Routing/test/UnitTests/RouteEndpointBuilderTest.cs:35]:
```suggestion
```

File: src/Http/Routing/test/UnitTests/Builder/RequestDelegateEndpointRouteBuilderExtensionsTest.cs
Comment (davidfowl) [inline src/Http/Routing/test/UnitTests/Builder/RequestDelegateEndpointRouteBuilderExtensionsTest.cs:234]: Breaking change?

File: src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs
Comment (Kahbazi) [inline src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:429]: ❓ 
```suggestion
            routeEndpointDataSource = new RouteEndpointDataSource(endpoints.ServiceProvider ?? EmptyServiceProvider.Instance, throwOnBadRequest);
```

**Comment by javiercn** [comment]
> Filters will no longer run for razor pages due to this change.

This might be problematic as it will break people's expectations that these filters run in both cases like other MVC filters. Is there a specific reason to prevent these filters from applying to pages?

File: src/Http/Routing/src/RouteEndpointDataSource.cs
Comment (captainsafia) [reply src/Http/Routing/src/RouteEndpointDataSource.cs:167]: I think adding it only for route handlers will actually help with our goal of making `WithOpenApi` something you can only apply on `RouteHandlerBuilder` types even though we removed that constraint. AFAIK, non-route handler scenarios don't need access to the method info.

File: src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs
Comment (captainsafia) [inline src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:205]: I may have missed it but where are we populating metadata from attributes now?

File: src/Http/Routing/src/RouteEndpointDataSource.cs
Comment (captainsafia) [inline src/Http/Routing/src/RouteEndpointDataSource.cs:278]: Nice! I like this.

File: src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs
Comment (halter73) [reply src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:205]: `RouteEndpointDataSource` now takes care of that bit since this removed the simpler `ModelEndpointDataSource` which would allow group conventions to unintentionally override endpoint-specific metadata like attributes.

Of course, when conventions are aware of groups, they might want to actually see this metadata as you pointed out #42750. The silver lining here is that these endpoints were never eligible for `.WithOpenApi()` treatment to begin with due to the lack of `MethodInfo` in the metadata, and this much remains unchanged so we don't make that issue any worse. And we're going to fix that issue hopefully anyway.

File: src/Http/Routing/src/RouteEndpointDataSource.cs
Comment (halter73) [reply src/Http/Routing/src/RouteEndpointDataSource.cs:167]: That's the question. Do we want OpenAPI for these endpoints or not? What if someone manually adds `[Consumes]` or `[Produces]` metadata to these endpoints? For now, I'm trying to keep this behavior consistent with the behavior we've already have.

**Comment by halter73** [comment]
> > Filters will no longer run for razor pages due to this change.
> 
> This might be problematic as it will break people's expectations that these filters run in both cases like other MVC filters. Is there a specific reason to prevent these filters from applying to pages?

This is my position as well. The problem is the way they were running in earlier previews is over the `RequestDelegate` (i.e. `Func<HttpContext, Task>` for the `Endpoint`s create for razor pages, and did not provide full fidelity `MethodInfo`s or invocation contexts the way other MVC filters do.

The same is true right now for MVC actions (until after this change merged after which those too will be skipped). @davidfowl opened a draft PR (#42557) to run these filters the way our customers expect for MVC actions. I'm going to take this over since he's on vacation.

We wanted to include razor pages in this PR as well, but the problem is that razor pages can effectively have multiple `MethodInfo`s per endpoint, and that doesn't fit well with our new filter abstraction. We hope to resolve this issue and run filters razor pages in a future release, but we don't know that we can do that in time for .NET 7. We've been considering whether we can "light up" filters for razor pages in future releases without breaking apps.

File: src/Http/Routing/test/UnitTests/Builder/RequestDelegateEndpointRouteBuilderExtensionsTest.cs
Comment (halter73) [reply src/Http/Routing/test/UnitTests/Builder/RequestDelegateEndpointRouteBuilderExtensionsTest.cs:234]: Technically.

You'd had to have defined a custom attribute that implemented `IHttpMethodMetadata` which [no one on grepp.app has](https://grep.app/search?q=IHttpMethodMetadata) other than us in this very test, [in one other test](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/test/testassets/RoutingSandbox/UseEndpointRoutingStartup.cs#L80) which notably didn't need to be changed because it wasn't overriding anything, and in .NET 6 preview diffs before I removed it from the MVC attributes. We certainly don't ship any such attributes.

Plus, we already made this "breaking" change for route handlers between .NET 6 and .NET 7 with #40926. See https://github.com/dotnet/aspnetcore/pull/40926/files#diff-f6727c82b3f6cff19c30c9c793f3e47c1c2ff2035a6453198ff4cedad8e84557

I almost went the other way and had `RouteEndopintDataSource` copy the old .NET 6 behavior here so we didn't "break" anything for either `Delegate` or `RequestDelegate`, but it makes sense to configure metadata added implicitly by the `MapXxx` call like the `MethodInfo` and `HttpMethodMetadata` before any additional metadata configured by the user.

File: src/Http/Routing/src/RouteEndpointDataSource.cs
Comment (davidfowl) [reply src/Http/Routing/src/RouteEndpointDataSource.cs:167]: I think we should unify these endpoints mostly because of the amount of bugs where people end up calling the RequestDelegate overload and it things stop working.

**Comment by halter73** [comment]
I updated this PR to do the renames suggested in #43000 and #42592 which I assigned both to myself.

I also reacted to @davidfowl's and @captainsafia's feedback about adding the `MethodInfo` to the endpoint metadata for old-fashioned `RequestDelegate`-defined endpoints. This will make it show up both with in the `.WithOpenApi()` output and the ApiExplorer stuff we shipped in .NET 6 used by things like swashbuckle and nswag.

Are we sure want this to affect ApiExplorer too? I like that this keeps things consistent with `.WithOpenApi()` but we'll certainly want to make a breaking change announcement.

**Comment by halter73** [comment]
> Are we sure want this to affect ApiExplorer too?

Even if we do agree this is the direction we should take, I think we need an easy way to opt out of this globally for people who might not want this behavior but want to easily upgrade their existing apps using something like swashbuckle or nswag to .NET 7.

Applications might call into libraries that call `MapXXX` internally making adding metadata to these endpoints to the description might be tricky. Doing something like changing `app.MapLibraryEndpoints()` to `app.MapGroup("").IgnoreApi().MapLibraryEndpoints()` if an `IgnoreApi()` extension method existed. If .MapLibraryEndpoints() already returns an `IEndpointConventionBuilder`, you wouldn't even need the group with an empty prefix.

We might want to reopen https://github.com/dotnet/aspnetcore/issues/34068 and add an `IgnoreApi()` extension method that works for both ApiExplorer and WithOpenApi(). @martincostello

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (davidfowl) [inline src/Http/Routing/src/RouteEndpointBuilder.cs:23]: See https://github.com/dotnet/aspnetcore/pull/42557, this is being moved to the base and the name should be changed. This name is redundant.

Comment (halter73) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:23]: I did see that. I figured that was just you prototyping in that PR though. Where was the discussion for this? Is there an API proposal? This wasn't mentioned in #42592 in the weeks following its approval.

I agree that removing `RouteHandler` from the name and just calling it `FilterFactories` is better. I do not think this should be moved to base type unless we also add the `RoutePattern` to the base type as well which I don't see happening because of the shadowing. See the following:

https://github.com/dotnet/aspnetcore/blob/8fe74e4d3ef2d26f58e3a2dfef90c64233941cfe/src/Http/Http.Extensions/src/RequestDelegateFactoryOptions.cs#L55-L57

I know you think this creates too tight of a coupling between routing and RDF, but itis very limiting for filters to not have access to the route pattern. `EndpointBuilders` without a `RoutePattern` [don't really exist](https://grep.app/search?q=%3A%20EndpointBuilder&filter[lang][0]=C%23). They're all `RouteEndpointBuilders` because who has time to reimplement route matching without `RoutePatterns`?

Maybe we should unseal `RouteEndpointBuilder`. It seems that this would help [the one non-test, non-inert] `EndpointBuilder`, [ServiceComposer's CompositionEndpointBuilder](https://github.com/ServiceComposer/ServiceComposer.AspNetCore/blob/9eba39a0ab7c7a94751ba6b3e065d5558bf57510/src/ServiceComposer.AspNetCore/CompositionEndpointBuilder.cs) which really want to be a `RouteEndpointBuilder` considering it has a `RoutePattern` it cannot expose and it returns a `RouteEndpoint` from `Build()`. @mauroservienti

Comment (halter73) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:23]: When I made that comment in RequestDelegateFactoryOptions and when leaving my last comment minutes ago, I forgot that the direction of the Routing -> Http.Extension dependency makes it impossible to expose `RoutePattern`s to filters without some crazy type forwarding. That is sad. Maybe we'll just expose it via metadata 😆

I still think we should just unseal `RouteEndpointBuilder` and leave `FilterFactories` there.

Comment (davidfowl) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:23]: I don't see why we would leave it on the derived type when the type got renamed to Endpoint* and not Route*. It has nothing to do with routing at this point.

Comment (halter73) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:23]: Now that I realize that the RoutePattern cannot be exposed to the filters because of the layering, I feel less strongly about it being in RouteEndopintBuilder. I'll update the API proposals to match what's in https://github.com/dotnet/aspnetcore/pull/42557 before our normal Monday API review.

I wish the routing stuff was lower level though. I still haven't seen a use case for EndpointBuilders or Endpoints that are not a RouteEndopintBuilders or RouteEndpoints. Having to downcast all the time is annoying. Are there any examples of this? Or is it more of a just-in-case thing? I realize you could in theory do your own matching based on something other than the RoutePattern. I just haven't seen it.

File: src/Shared/RoslynUtils/TypeHelper.cs
Comment (halter73) [reply src/Shared/RoslynUtils/TypeHelper.cs:44]: This is shared source, but I think this is the only caller so I can make it private. I like calling the method over doing `Attribute.IsDefined(...) || IsCompilerGeneratedType(...)` inline, because I think it's a little more self-documenting.

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (captainsafia) [reply src/Http/Routing/src/RouteEndpointBuilder.cs:23]: The general use of `EndpointBuilder` as a context object makes sense to me. Is this use of builders-as-context-objects common anywhere else? The name can make it hard to understand that the builder is used by RDF and filters for accessing route properties, and not for actually modifying conventions on the route endpoint builder.

File: src/Http/Routing/src/ConfigureRouteHandlerOptions.cs
Comment (captainsafia) [inline src/Http/Routing/src/ConfigureRouteHandlerOptions.cs:13]: What's the motivation for this change?

Comment (halter73) [reply src/Http/Routing/src/ConfigureRouteHandlerOptions.cs:13]: It's to avoid breaking unit tests that add options to the service collection but not IHostEnvironment. There were a few tests in downstream projects like SignalR that were broken by this. This was a simple fix that might also avoid breaking external tests and odd configurations without an IHostEnvironment.

**Comment by halter73** [comment]
I think I've addressed all the PR feedback.

I undid the renames to reduce merge conflicts in #42557. The renames weren't the point of this PR. I thought I was saving work, but I wasn't. We also want to expose the `EndopintBuilder` in more places (see https://github.com/dotnet/aspnetcore/issues/43000#issuecomment-1201920724). This will be done in another PR.

File: src/Http/Routing/src/RouteEndpointBuilder.cs
Comment (halter73) [reply src/Http/Routing/src/RouteEndpointBuilder.cs]: > Is this use of builders-as-context-objects common anywhere else?

I'd consider it more of a builders-in-context-objects, but I don't know of another example in ASP.NET. Contexts and builders as separate parameters are more common, but that's for callbacks that are specifically intended to add to the builder.

It might actually make sense to change `IEndpointMetadataProvider.PopulateMetadata(EndpointMetadataContext)` to `IEndpointMetadataProvider.PopulateMetadata(MethodInfo, EndpointBuilder)` and `IEndpointParameterMetadataProvider.PopulateMetadata(EndpointParameterMetadataContext)` to `IEndpointParameterMetadataProvider.PopulateMetadata(ParameterInfo, EndpointBuilder)` and get rid of the new contexts entirely.

For filters, we could change the factory type from `Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>` to `Func<MethodInfo, EndopintBuilder, EndpointFilterDelegate, EndpointFilterDelegate>` and get rid of the `EndpointFilterFactoryContext`.

In the filter case, I like the flexibility of the context though, and the primary purpose of filters is probably not going to be to update metadata on the `EndpointBuilder`. This is the whole point of the meta data providers which is why I support removing the contexts there. @DamianEdwards 

> The name can make it hard to understand that the builder is used by RDF and filters for accessing route properties, and not for actually modifying conventions on the route endpoint builder.

I don't think this will be confusing because it's not like we're exposing an `IEndpointConventionBuilder` which has a bunch of extension methods targeting it that add conventions. The `EndpointBuilder` does make it easy to modify metadata which many conventions do, but filters and metadata providers are still going to have to add the metadata manually rather than use extension methods just like today.

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**
